File: generate.ts
#!/usr/bin/env -S bun run

import { opendir } from "node:fs/promises";

import path from "node:path";
import process from "node:process";

import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkGfm from "remark-gfm";
import remarkMath from "remark-math";
import remarkRehype from "remark-rehype";
import rehypeRaw from "rehype-raw";
import rehypeKatex from "rehype-katex";
import rehypeHighlight from "rehype-highlight";
import rehypeExternalLinks from "rehype-external-links";
import rehypeFormat from "rehype-format";
import rehypeStringify from "rehype-stringify";
import { visit } from "unist-util-visit";
import remarkFrontmatter from "remark-frontmatter";
import { matter } from "vfile-matter";

import type { ElementContent, Root } from "hast";

import * as cheerio from "cheerio";

import { $ } from "bun";

process.chdir(import.meta.dir);

const website_url = "https://beaumccartney.com";

const src_dir = "src";
const build_dir = "build";
const css_dir = "css";
const css_out_dir = path.join(build_dir, css_dir);
const fonts_out_dir = path.join(css_out_dir, "fonts");
const blog_dir = "blog";
const katex_css_out = "katex.css";
const highlight_css_out = "highlight.css";
const assets_dir = "assets";
const markdown_extension = ".md";
const rss_file = "rss.xml";

await $`rm -rf ${build_dir}`;
await $`mkdir -p ${fonts_out_dir}`;
await $`cp -r ${assets_dir}/* build`;

const node_modules_dir = "node_modules/";
await $`cp ${path.join(node_modules_dir, "highlight.js/styles/base16/gruvbox-dark-medium.min.css")} ${path.join(css_out_dir, highlight_css_out)}`;

const katex_module = path.join(node_modules_dir, "katex/dist");
await $`cp ${path.join(katex_module, "katex.min.css")} ${path.join(css_out_dir, katex_css_out)}`;
await $`cp ${path.join(katex_module, "fonts")}/* ${fonts_out_dir}`;

const complete_page_template = (
  title: string,
  content: string,
  description: string,
  url: string,
  og_type: string,
  og_extras: { property: string; content: string }[],
) => `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>${title}</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <meta name="author" content="Beau McCartney">
    <meta name="description" content="${description}">
    <meta
      name="keywords"
      content="personal, portfolio, website, beau, mccartney, Beau McCartney, software, engineer, computer, graphics, gamedev"
    >
    <meta name="robots" content="index, follow">
    <meta property="og:title" content="${title}">
    <meta property="og:url" content="${url}">
    <meta property="og:type" content="${og_type}">
    <meta property="og:description" content="${description}">
    ${og_extras.map((extra) => `<meta property="${extra.property}" content="${extra.content}">`).join("\n")}

    <link rel="stylesheet" href="${path.join("/", css_dir, katex_css_out)}">

    <link rel="stylesheet" href="${path.join("/", css_dir, "sakura-dark.css")}">

    <link rel="stylesheet" href="${path.join("/", css_dir, highlight_css_out)}">

    <link rel="alternate" type="application/rss+xml" title="RSS" href="/${rss_file}">
    <script data-goatcounter="https://beaumccartney.goatcounter.com/count" async src="//gc.zgo.at/count.js"></script>
  </head>
  <body>
    <main>
    ${content}
    </main>
  </body>
</html>`;

// helper for computing the footnote number in the template
function footnote_number(ix: number): number {
  return ix + 1;
}

const footnote_link_id_template = (ix: number) => `fn_link-${ix}`;

const footnote_id_template = (ix: number) => `fn-${ix}`;

const time_element_template = (thedate: string) =>
  `<time datetime="${thedate}">${thedate}</time>`;

type Page = {
  content: string;
  title: string;
  description: string;
  publish_date: string | null;
};
async function process_markdown_content(file: Bun.BunFile): Promise<Page> {
  const markdown = await file.text();

  const file_out = await unified()
    .use(remarkParse)
    .use(remarkFrontmatter, ['yaml'])
    .use(() => (_, file_v) => { matter(file_v); })
    .use(remarkGfm, { singleTilde: false })
    .use(remarkMath)
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(rehypeRaw)
    .use(rehypeKatex)
    .use(rehypeExternalLinks, { target: "_blank", rel: ["nofollow", "noopener", "noreferrer"] })
    .use(rehypeHighlight)
    .use(() => function(tree: Root) {
      const notes: ElementContent[][] = [];
      visit(tree, "element", (node, index, parent) => {
        if (node.tagName === "fn" && parent && typeof index === "number") {
          const ix = notes.length;
          const note_children = node.children;
          notes.push(note_children);
          parent.children[index] = {
            type: "element",
            tagName: "sup",
            properties: { id: footnote_link_id_template(ix) },
            children: [
              {
                type: "element",
                tagName: "a",
                properties: { href: `#${footnote_id_template(ix)}` },
                children: [{ type: "text", value: String(footnote_number(ix)) }],
              },
            ],
          };
        }
      });
      if (notes.length) {
        tree.children.push({ type: "element", tagName: "hr", properties: {}, children: [] });
        const ol_children: ElementContent[] = notes.map((note, ix) => ({
          type: "element",
          tagName: "li",
          properties: { id: footnote_id_template(ix) },
          children: [
            {
              type: "element",
              tagName: "p",
              properties: {},
              children: [
                {
                  type: "element",
                  tagName: "sup",
                  properties: {},
                  children: [
                    {
                      type: "element",
                      tagName: "a",
                      properties: { href: `#${footnote_link_id_template(ix)}` },
                      children: [{ type: "text", value: `${footnote_number(ix)}.` }],
                    },
                  ],
                },
                { type: "text", value: " " },
                ...note,
              ],
            },
          ],
        }));
        tree.children.push({
          type: "element",
          tagName: "ol",
          properties: { style: "list-style-type: none; padding-left: 0;" },
          children: ol_children,
        });
      }
    })
    .use(rehypeFormat)
    .use(rehypeStringify, { allowDangerousHtml: true })
    .process(markdown);

  const html = String(file_out);
  const $page = cheerio.load(html, {}, false);

  const fm = file_out.data.matter as Record<string, unknown> | undefined;
  const description = fm?.description;
  if (!description || typeof description !== "string") {
    throw new Error(`page ${file.name} missing description frontmatter`);
  }
  const publish_date = fm?.publish_date;
  if (publish_date !== undefined && typeof publish_date !== "string") {
    throw new Error(`page ${file.name} gave invalid publish date`);
  }

  return {
    content: html,
    title: $page("h1").text(),
    description,
    publish_date: publish_date ?? null,
  };
}

type Post = {
  title: string;
  folder: string;
  url: string;
  publish_date: string;
  description: string;
};
const posts: Post[] = [];
for await (const dirent of await opendir(path.join(src_dir, blog_dir))) {
  if (!dirent.isFile() || path.extname(dirent.name) != markdown_extension) {
    throw new Error("non-markdown entry in blog");
  }

  const post_content = await process_markdown_content(
    Bun.file(path.join(dirent.parentPath, dirent.name)),
  );
  const $post = cheerio.load(
    `<article>\n${post_content.content}\n</article>`,
    {},
    false,
  );

  const publish_date = post_content.publish_date;
  if (!publish_date) {
    throw new Error(`post [${post_content.title}] has no publish date`);
  }
  $post("h1").after(
    `<p><strong>${time_element_template(publish_date)}</strong></p>`,
  );
  const folder = path.join(
    blog_dir,
    dirent.name.slice(0, -markdown_extension.length),
  );
  const post: Post = {
    title: post_content.title,
    folder,
    url: path.join(website_url, folder),
    publish_date,
    description: post_content.description,
  };
  posts.push(post);

  const post_page = complete_page_template(
    post.title,
    $post.html(),
    post.description,
    post.url,
    "article",
    [
      {
        property: "article:published_time",
        content: post.publish_date,
      },
    ],
  );

  Bun.write(path.join(build_dir, folder, "index.html"), post_page);
}

{
  const home = await process_markdown_content(
    Bun.file(path.join(src_dir, `index${markdown_extension}`)),
  );
  posts.sort((a, b) => b.publish_date.localeCompare(a.publish_date));
  const $home = cheerio.load(home.content, {}, false);
  $home("#blog-entries").html(`<ul>
  ${posts.map((post) => `<li>${time_element_template(post.publish_date)} - <a href="${path.join("/", post.folder)}">${post.title}</a></li>`).join("\n")}
</ul>
<a href="/rss.xml">RSS Feed</a>`);

  const home_page = complete_page_template(
    home.title,
    $home.html(),
    home.description,
    website_url,
    "profile",
    [
      {
        property: "profile:first_name",
        content: "Beau",
      },
      {
        property: "profile:last_name",
        content: "McCartney",
      },
    ],
  );

  Bun.write(path.join(build_dir, "index.html"), home_page);
}

const rss_feed = `<?xml version="1.0" encoding="UTF-8" ?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">

<channel>
  <title>Beau McCartney</title>
  <link>${website_url}</link>
  <description>Beau McCartney's website and blog</description>
  <language>en-us</language>
  <atom:link href="${website_url}/rss.xml" rel="self" type="application/rss+xml" />
  ${posts
    .map((post) => {
      const date = new Date(post.publish_date);
      const rss_publish_date = date.toUTCString();

      return `<item>
  <title>${post.title}</title>
  <link>${post.url}</link>
  <guid isPermaLink="true">${post.url}</guid>
  <pubDate>${rss_publish_date}</pubDate>
  <description>${post.description}</description>
</item>`;
    })
    .join("\n")}
</channel>
</rss>`;

Bun.write(path.join(build_dir, rss_file), rss_feed);

File: generate.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"log"
	"os"
	"slices"
	"strings"

	"time"

	"io/fs"
	"path/filepath"

	"text/template"

	"github.com/yuin/goldmark"
	"github.com/yuin/goldmark/extension"
	"github.com/yuin/goldmark/parser"
	md_html "github.com/yuin/goldmark/renderer/html"

	"github.com/FurqanSoftware/goldmark-katex"
	"go.abhg.dev/goldmark/frontmatter"

	"github.com/PuerkitoBio/goquery"
)

const (
	WEBSITE_URL = "https://beaumccartney.com"

	BUILD_RELATIVE_TO_SRC_PATH = "../build/"
	SRC_PATH = "src/"
	BLOG_PATH = "blog/"

	RSS_FILE = "rss.xml"
	MD_EXTENSION = ".md"

	FOOTNOTE_PLACEHOLDER_ID = "footnotes-placeholder"
	FOOTNOTES_PLACEHOLDER = `<div id="` + FOOTNOTE_PLACEHOLDER_ID +`"></div>`
)

type Post struct {
	Title string
	Relative_url string
	Description string
	Pubdate string
	Html bytes.Buffer
}

func absolute_url(relative_url string) string {
	result := WEBSITE_URL
	if len(relative_url) > 0 && relative_url[0] != '/' {
		result += "/"
	}
	result += relative_url

	return result
}

// NOTE(beau): define indexing convention for footnote superscripts
func footnote_number(ix int) int { return ix + 1 }
func footnote_link_target_id(ix int) string {
	return fmt.Sprintf("fn-%d", ix)
}
func footnote_backlink_target_id(ix int) string {
	return fmt.Sprintf("fn_backlink_target-%d", ix)
}

func main() {
	md := goldmark.New(
		goldmark.WithRendererOptions(
			md_html.WithUnsafe(),
		),
		goldmark.WithExtensions(
			extension.Strikethrough,
			&frontmatter.Extender{},
			&katex.Extender{},
		),
	)
	os.Chdir(SRC_PATH)

	var (
		posts []Post
		index_page_html bytes.Buffer
	)

	err := filepath.WalkDir(".", func(src_relative_path string, d fs.DirEntry, err error) error {
		if d.Type().IsRegular() {
			if !strings.HasSuffix(src_relative_path, MD_EXTENSION) {
				dst_path := filepath.Join(BUILD_RELATIVE_TO_SRC_PATH, src_relative_path)

				var (
					src_file, dst_file *os.File
					err error
				)

				if src_file, err = os.Open(src_relative_path); err != nil {
					log.Fatal(err)
				}

				if err = os.MkdirAll(filepath.Dir(dst_path), 0755); err != nil {
					log.Fatal(err)
				}

				dst_file, err = os.Create(dst_path)


				if _, err := io.Copy(src_file, dst_file); err != nil {
					log.Fatal(err)
				}

				return nil
			}

			md_path := src_relative_path
			src_md, err := os.ReadFile(md_path)
			if err != nil {
				log.Fatalf("couldn't open [%s]\n", md_path)
			}

			var (
				converted_html bytes.Buffer
				fm map[string]string
			)
			{
				ctx := parser.NewContext()
				if err = md.Convert(src_md, &converted_html, parser.WithContext(ctx)); err != nil {
					log.Fatalf("couldn't convert [%s] to html", md_path)
				}
				d := frontmatter.Get(ctx)
				if d == nil {
					log.Fatal("no frontmatter")
				}
				if err = d.Decode(&fm); err != nil {
					log.Fatal(err)
				}
				fmt.Printf("converted for [%s]\n\n%s\n\n\n\n", md_path, &converted_html)
			}

			var (
				page_content string
				post Post
				title string
			)
			{
				doc, err := goquery.NewDocumentFromReader(bytes.NewReader(page_html))
				if err != nil {
					log.Fatal(err)
				}

				{
					h1 := doc.Find("h1")
					title = h1.Text()
					h1.Remove()
				}
			}

			description := fm["description"]
			var page_html bytes.Buffer
			isblog := strings.HasPrefix(md_path, BLOG_PATH)
			isindex := md_path == "index" + MD_EXTENSION
			if isblog {
				pubdate := fm["publish_date"]

				filename := filepath.Base(md_path)
				filename_no_ext := filename[:len(filename)-len(filepath.Ext(filename))]
				dst_relative_path := filepath.Join(BLOG_PATH, filename_no_ext, "index.html")
				post = Post{
					Title: title,
					Relative_url: dst_relative_path,
					Description: description,
					Pubdate: pubdate,
				}
				if err = blog_page(
					page_content,
					title,
					description,
					dst_relative_path,
					pubdate,
				).Render(context.Background(), &page_html); err != nil {
					log.Fatal(err)
				}
			} else if isindex {
				err = index_page(
					page_content,
					title,
					description,
				).Render(context.Background(), &page_html)
			} else {
				log.Fatal("markdown file that's not blog or index encountered")
			}
			if err != nil {
				log.Fatalf("couldn't render page template for [%s]", md_path)
			}

			var html_with_footnotes string
			{
				doc, err := goquery.NewDocumentFromReader(&page_html)
				h1 := doc.Find("h1")
				h1.Remove()

				var footnotes []string
				doc.Find("fn").Each(func(ix int, s *goquery.Selection) {
					note, err := s.Html()
					if err != nil { log.Fatal(err) }
					footnotes = append(footnotes, note)
					fn_link := fmt.Sprintf(
						"<sup id=\"%s\"><a href=\"#%s\">%d</a></sup>",
						footnote_backlink_target_id(ix),
						footnote_link_target_id(ix),
						footnote_number(ix),
					)
					s.ReplaceWithHtml(fn_link)
				})

				err = footnote_block(footnotes).Render(
					context.Background(),
					strings.NewWriter(bytes.NewBufferString(html_with_footnotes)),
				)
				if page_content, err = doc.Html(); err != nil {
					log.Fatal(err)
				}
				fmt.Printf("content for [%s]\n\n%s\n\n\n\n", md_path, page_content)
			}

			if isblog {
				post.Html = html_with_footnotes
				posts = append(posts, post)
			} else if isindex {
				index_page_html = html_with_footnotes
			}
		}

		return nil
	})
	if err != nil {
		log.Fatal("error with walking files", err)
	}


	// write index page
	{
		var doc *goquery.Document
		doc, err = goquery.NewDocumentFromReader(&index_page_html)
		if err != nil {
			log.Fatal(err)
		}
		slices.SortFunc(posts, func(a Post, b Post) int {
			result := 0
			if a.Pubdate < b.Pubdate {
				result = 1
			} else if a.Pubdate > b.Pubdate {
				result = -1
			}
			return result
		})
		var posts_index_html bytes.Buffer
		posts_index(posts).Render(context.Background(), &posts_index_html)
		doc.Find("#blog-entries").ReplaceWithHtml(posts_index_html.String())

		var final_index_html string
		if final_index_html, err = doc.Html(); err != nil {
			log.Fatal(err)
		}
		if err = os.MkdirAll(BUILD_RELATIVE_TO_SRC_PATH, 0755); err != nil {
			log.Fatal(err)
		}
		if err = os.WriteFile(
			filepath.Join(BUILD_RELATIVE_TO_SRC_PATH, "index.html"),
			[]byte(final_index_html),
			0644,
		); err != nil {
			log.Fatal(err)
		}
	}

	for _, post := range posts {
		dst_path := filepath.Join(BUILD_RELATIVE_TO_SRC_PATH, post.Relative_url)
		if err = os.MkdirAll(filepath.Dir(dst_path), 0755); err != nil {
			log.Fatal(err)
		}
		if err = os.WriteFile(
			dst_path,
			post.Html.Bytes(),
			0644,
		); err != nil {
			log.Fatal(err)
		}
	}

	// rss feed
	{
		rss_info := struct {
			Website_url string
			Rss_file string
			Posts []Post
		}{
			Website_url: WEBSITE_URL,
			Rss_file: RSS_FILE,
			Posts: slices.Clone(posts),
		}
		for ix := range rss_info.Posts {
			post := &rss_info.Posts[ix]
			// t, err := time.Parse(time.DateOnly, post.Pubdate)
			t, err := time.Parse(
				time.DateOnly + " 15:04 MST", // support automatic 12pm mst pub time
				post.Pubdate + " 12:00 MST",  // 12pm mst pub time
			)
			if err != nil {
				log.Fatal(err)
			}
			post.Pubdate = t.Format("Mon, 02 Jan 2006 15:04:05 MST")
		}
		tmpl, err := template.New("rss").Funcs(template.FuncMap{ "absolute_url": absolute_url }).Parse(`<?xml version="1.0" encoding="UTF-8" ?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
<channel>
  <title>Beau McCartney</title>
  <link>{{ .Website_url }}</link>
  <description>Beau McCartney's website and blog</description>
  <language>en-us</language>
  <atom:link href="{{ absolute_url .Rss_file }}" rel="self" type="application/rss+xml" />
  {{ range .Posts }}
  <item>
      <title>{{ .Title }}</title>
      <link>{{ absolute_url .Relative_url }}</link>
      <guid isPermaLink="true">{{ absolute_url .Relative_url }}</guid>
      <pubDate>{{ .Pubdate }}</pubDate>
      <description>{{ .Description }}</description>
    </item>
  {{ end }}
</channel>
</rss>`)
		var rss_file *os.File
		if rss_file, err = os.Create(filepath.Join(BUILD_RELATIVE_TO_SRC_PATH, RSS_FILE)); err != nil {
			log.Fatal(err)
		}
		if err = tmpl.Execute(rss_file, rss_info); err != nil {
			log.Fatal(err)
		}
	}
}
